### PR TITLE
T122: Plan form backend

### PR DESCRIPTION
**Pull request by kdata12**
## Description 

Created an api endpoint for handling plan form submission.

## Types of changes

What types of changes does your code introduce to WeWrap?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

- [x] I have made sure that eslint was passed locally
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by kdata12** [comment]
You can ignore the frontend files, I just checked them out from `hyun/plan-form-frontend` to test the route!

File: backend/routes/planForm.ts
Comment (justinpchen94) [inline backend/routes/planForm.ts:1]: let's make this a rule!

File: backend/routes/planForm.ts
Comment (justinpchen94) [inline backend/routes/planForm.ts]: is this usage correct? this feels like it's creating a tonnnn of validation middlewares

FWIW I've used `express-validator` in conjunction with the `joi` package in the past and this is pretty standard in case you're looking for an alternative

File: backend/routes/planForm.ts
Comment (justinpchen94) [inline backend/routes/planForm.ts]: is 400 the right code here? we technically don't know if an error here represents a bad input or some other 500 error right?

**Review by justinpchen94 — COMMENTED** [review]
some small nits but will wait until frontend is merged to main and you rebase this branch to get rid of the frontend diffs before approving

File: backend/routes/planForm.ts
Comment (kdata12) [reply backend/routes/planForm.ts]: Added error handling as a todo task

File: backend/routes/planForm.ts
Comment (kdata12) [reply backend/routes/planForm.ts]: I believe that this is the right usage. Here's the doc: https://express-validator.github.io/docs/#basic-guide

Can you show me what you mean by using `express-validator` with `joi`. Also, I'm new to `express-validator`.

Comment (justinpchen94) [reply backend/routes/planForm.ts]: try checking some examples out like this https://www.npmjs.com/package/express-joi-validation

File: backend/routes/planForm.ts
Comment (kdata12) [reply backend/routes/planForm.ts]: Can I just keep the error there as a placeholder until I actually implement proper error handling?

File: backend/app.ts
Comment (justinpchen94) [inline backend/app.ts]: prob just disable this rule

File: backend/routes/planForm.ts
Comment (justinpchen94) [inline backend/routes/planForm.ts]: so in general we don't usually want a route file like this to contain the code that directly modifies the db. a standard name for this is called a "3 layer service architecture" 

see this article: https://blog.devgenius.io/the-three-layer-architecture-for-node-js-applications-ce32a3a30fa6

the names are relatively interchangeable but typically I see/use "routes" "controllers" and "models" as the main 3 layers. 

In this route file, we receive and validate the request schema and in the request handler we are passed the `req` and `res`. generally we never pass the `req` and `res` further down and leave it in this file. 

we pull out the values we need, and pass them into `controller` methods. in this example i'd expect there to be a `backend/controllers/plan.ts` file of some sort and it would have a method called like `createPlan` that takes in the values in your `req.body`.  generally all of the true business logic lives inside the controller file

And then finally we have `model` files, which would house all the code that directly calls the prisma client and calls the actual db methods. all files wishing to make db updates would then be able to go through methods in this file, rather than having db calls just spread out everywhere

File: backend/routes/planForm.ts
Comment (justinpchen94) [reply backend/routes/planForm.ts]: as long as you add a todo!

**Review by justinpchen94 — APPROVED** [review]
lgtm to unblock but you could do the refactor in a different pr if you'd like

File: backend/service/planMembership.ts
Comment (kdata12) [inline backend/service/planMembership.ts]: `const friendsPlanMembership` is not working at the moment because we haven't set up the frontend to send a the proper array of users that prisma wants

File: backend/models/plan.ts
Comment (kdata12) [inline backend/models/plan.ts:9]: For this data, should we be more explicit and readable by typing `incoming.description`, `incoming.startDate` and etc. Or should we do `data: incomingData` here

File: backend/app.ts
Comment (justinpchen94) [reply backend/app.ts]: can we rm this comment if the rule has been turned off?

File: backend/controllers/plan.ts
Comment (kdata12) [inline backend/controllers/plan.ts:22]: placeholder error handlers

File: backend/app.ts
Comment (justinpchen94) [inline backend/app.ts]: nit: in following with REST naming conventions, this should just be `/api/plan`

File: backend/controllers/plan.ts
Comment (justinpchen94) [reply backend/controllers/plan.ts:22]: do you have to check for null here as well or is the expected return `undefined`?

File: backend/models/plan.ts
Comment (justinpchen94) [reply backend/models/plan.ts:9]: honestly when in doubt, more explicit the better is generally the rule of thumb

File: backend/models/plan.ts
Comment (justinpchen94) [inline backend/models/plan.ts:17]: can just do `if (responseData)` i believe

File: backend/models/plan.ts
Comment (justinpchen94) [inline backend/models/plan.ts]: should be a `console.error`

File: backend/models/planMembership.ts
Comment (justinpchen94) [inline backend/models/planMembership.ts]: same - `console.error`

File: backend/models/planMembership.ts
Comment (justinpchen94) [inline backend/models/planMembership.ts]: `console.error`

File: backend/modules/auth.ts
Comment (justinpchen94) [inline backend/modules/auth.ts]: will this handle a redirect?

**Review by justinpchen94 — APPROVED** [review]
mostly small nits, looking good!

File: backend/controllers/plan.ts
Comment (kdata12) [reply backend/controllers/plan.ts:22]: For the prisma create methods, they either return the data or undefined.

File: backend/modules/auth.ts
Comment (kdata12) [reply backend/modules/auth.ts]: yeah i think it should redirect the user to an error page

File: backend/models/plan.ts
Comment (kdata12) [reply backend/models/plan.ts:17]: this would throw an error when responseData has been fulfilled

File: backend/modules/auth.ts
Comment (kdata12) [reply backend/modules/auth.ts]: right now, this is just going to redirect to the main page as a placehodler. Added as a task

**Pull request merged**
